### PR TITLE
Allows defining multiple compoenents of a specific type on a single e…

### PR DESCRIPTION
**Pull request by dmarcos**

### CODE REVIEW

File: src/core/a-entity.js [hunk 1]
```diff
@@ -290,19 +294,23 @@ var proto = Object.create(ANode.prototype, {
 
       // Check if component already initialized.
       if (name in this.components) { return; }
-      component = this.components[name] = new components[name].Component(this, data);
+      if (id && !components[name].multiple) {
+        throw new Error('Trying to initialize multiple components of type ' + name +
```
Comment (ngokevin) [inline src/core/a-entity.js]: can we indent like we do here: https://github.com/aframevr/aframe/pull/1596/files#diff-dc3043915bc0e1dc643b7fa488e01c0eR254

Comment (ngokevin) [inline src/core/a-entity.js]: and put backticks around `name`

File: src/core/a-entity.js [hunk 2]
```diff
@@ -278,7 +278,11 @@ var proto = Object.create(ANode.prototype, {
   initComponent: {
     value: function (name, data, isDependency) {
       var component;
+      var componentInfo = name.split('__');
+      var id = componentInfo[1];
```
Comment (ngokevin) [inline src/core/a-entity.js]: `componentId` or `instanceId` maybe to diff from HTML ID

File: src/components/sound.js [hunk 3]
```diff
@@ -41,16 +42,28 @@ module.exports.Component = registerComponent('sound', {
...
-      el.addEventListener(data.on, this.play.bind(this));
+      this.updateEventListener(oldData.on);
     }
 
     // All sound values set. Load in `src`.
     if (srcChanged) { sound.load(data.src); }
   },
 
+  /**
+  *  update listener attached to the user defined on event.
```
Comment (ngokevin) [inline src/components/sound.js]: nit: could cap this sentence

File: tests/core/a-entity.test.js [hunk 4]
```diff
@@ -549,6 +571,24 @@ suite('a-entity', function () {
...
+
+    test('does not initialize with id if the component is not multiple', function () {
+      var el = this.el;
+      assert.throws(function setAttribute () {
+        el.setAttribute('geometry__1', {primitive: 'box'});
+      }, Error);
+      assert.notOk(el.components.geometry_1);
+    });
+
+    test('initializes components with id if it opts in to multiple', function () {
```
Comment (ngokevin) [inline tests/core/a-entity.test.js]: nit: `s/in to/into` and `s/it/the component`

File: tests/core/a-entity.test.js [hunk 5]
```diff
@@ -381,6 +393,16 @@ suite('a-entity', function () {
       el.setAttribute('class', 'pied piper');
       assert.equal(el.getAttribute('class'), 'pied piper');
     });
+
+    test('retrieves data from a multiple component', function () {
```
Comment (ngokevin) [inline tests/core/a-entity.test.js]: could probably test `getComputedAttribute` as well

Comment (ngokevin) [inline tests/core/a-entity.test.js]: we should also test multi-component detach

File: tests/components/sound.test.js [hunk 6]
```diff
@@ -75,4 +71,19 @@ suite('sound', function () {
       assert.ok(sound.pause.called);
     });
   });
+
+  suite('play', function () {
```
Comment (ngokevin) [inline tests/components/sound.test.js]: could test the sound component if used with multi, that the object3dname is set correctly

**Comment by ngokevin** [comment]
I found `removeAttribute` doesn't detach the component.

File: tests/components/sound.test.js (hunk 6 above)
Comment (dmarcos) [inline tests/components/sound.test.js]: object3Dname uses now this.attrName so multi or not it's the same code path

Comment (ngokevin) [inline tests/components/sound.test.js]: maybe you can test that `this.attrName` is correctly being used with `setObject3D`

File: src/core/a-entity.js [hunk 7]
```diff
@@ -271,38 +271,48 @@ var proto = Object.create(ANode.prototype, {
...
+   * @param {string} attrName - Attribute name asociated to the component.
    * @param {object} data - Component data
    * @param {boolean} isDependency - True if the component is a dependency.
    */
   initComponent: {
-    value: function (name, data, isDependency) {
+    value: function (attrName, data, isDependency) {
       var component;
-      var isComponentDefined = checkComponentDefined(this, name) || data !== undefined;
+      var componentInfo = attrName.split('__');
```
Comment (ngokevin) [inline src/core/a-entity.js]: Probably good to make a constant for this.

`MULTI_DELIMITER = '__'`;

File: src/core/a-entity.js [hunk 8]
```diff
@@ -271,38 +271,48 @@ var proto = Object.create(ANode.prototype, {
...
+      this.initComponentDependencies(componentName);
 
-      // Check if component already initialized.
-      if (name in this.components) { return; }
-      component = this.components[name] = new components[name].Component(this, data);
+      // If component name has an id we check component type multiplic
+      if (componentId && !components[componentName].multiple) {
+        throw new Error('Trying to initialize multiple ' +
+                        'components of type `' + componentName +
+                        '` There can only be one component of this type per entity.');
```
Comment (ngokevin) [inline src/core/a-entity.js]: period before There

**Comment by ngokevin** [comment]
Add test for `getComputedAttribute` for multi

r+wc

**Comment by ngokevin** [comment]
nice

**Pull request merged**